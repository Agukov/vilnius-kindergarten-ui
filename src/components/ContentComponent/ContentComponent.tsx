import * as React from 'react';
import './content-component.css';
import { Table } from 'reactstrap';
import { Link } from 'react-router-dom';
import axios from 'axios';

export interface ContentComponentProps {
  kindergartens: Kindergarten[];
}

export interface Kindergarten {
  id: number;
  label: string;
  address: string;
  buildDate: string;
  elderate: string;
  email: string;
  phone: string;
  schoolType: string;
  www: string;
  idFromSource: number;
}

export class ContentComponent extends React.Component<ContentComponentProps, {gardens: Kindergarten[]}> {
  constructor(props: ContentComponentProps) {
    super(props);
    this.state = {gardens: []};
  }

  componentWillMount() {
    axios
    .get('https://safe-mesa-80356.herokuapp.com/api/garden')
    .then(res => this.setState({gardens: res.data}))
    .catch(err => {throw new ReferenceError('Failed to load gardens data'); });
  }

  renderKindergartenList() {
    // create rows in table
    const kindList = this.state.gardens.map((kin, idx, aaa) => {
      return (
        <tr key={idx}>
          <th scope="row">{idx + 1}</th>
          <td><Link to={`gardens/${kin.id}`}>{kin.label}</Link></td>
          <td>{kin.address}</td>
          <td>{kin.elderate}</td>
        </tr>
      );
    });
    // return tab
    return (
      <Table>
        <thead>
          <tr>
            <th>#</th>
            <th>Name</th>
            <th>Addres</th>
            <th>Type</th>
          </tr>
        </thead>
        <tbody>
          {kindList}
        </tbody>
      </Table>
    );
  }

  render() {
    return this.state.gardens.length === 0 ? <div>Loading...</div> : (
      <div>
        {this.renderKindergartenList()}
      </div>);
  }
}
