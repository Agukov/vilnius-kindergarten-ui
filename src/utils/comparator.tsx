import { Kindergarten } from '../components/ContentComponent/ContentComponent';

export const fromArray = {
  get3Max: (gardens: Kindergarten[]) => {
    const newGardens = gardens.concat();
    newGardens.sort((a, b) => {
      if (a.buildDate === b.buildDate) { return 0; }
      return a.buildDate > b.buildDate ? -1 : 1;
    });
    return newGardens
      .filter(g => g.www !== '')
      .slice(0, 3);
  }
};
