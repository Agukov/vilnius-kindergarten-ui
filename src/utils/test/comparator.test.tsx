import { fromArray } from '../comparator';
import { Kindergarten } from '../../components/ContentComponent/ContentComponent';
import * as moment from 'moment';

export const MOCK_GARDENS: Kindergarten[] = [
  {
    id: 1,
    label: 'string',
    address: 'string',
    buildDate: 1201355534609,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 2,
    label: 'string',
    address: 'string',
    buildDate: 1601355537609,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 3,
    label: 'string',
    address: 'string',
    buildDate: 1111357534604,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 4,
    label: 'string',
    address: 'string',
    buildDate: 1501357834678,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 5,
    label: 'string',
    address: 'string',
    buildDate: 1001355534654,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 6,
    label: 'string',
    address: 'string',
    buildDate: 1011445534677,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 7,
    label: 'string',
    address: 'string',
    buildDate: 1221355534611,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 8,
    label: 'string',
    address: 'string',
    buildDate: 902255534689,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: 'string',
    idFromSource: 2,
  },
  {
    id: 9,
    label: 'string',
    address: 'string',
    buildDate: 1821355534611,
    elderate: 'string',
    email: 'string',
    phone: 'string',
    schoolType: 'string',
    www: '',
    idFromSource: 2,
  }
];
expect.addSnapshotSerializer({
  test: (val: Kindergarten) => !!val.buildDate,
  print: (val: Kindergarten) => `id: ${val.id} build date: ${moment(val.buildDate).format()} www: ${val.www}`
});

describe('Compare functions', () => {
  // test with Snapshot testing
  it('should return 3 kidergardens with highset build date (ST)', () => {
    expect(MOCK_GARDENS).toMatchSnapshot();
    expect(fromArray.get3Max(MOCK_GARDENS)).toMatchSnapshot();
    expect(MOCK_GARDENS).toMatchSnapshot();
  });
  // classic test
  it('should return 3 kidergardens with highset build date', () => {
    const gardensBeforeSort = MOCK_GARDENS.map(g => g.id).join(' ');
    // console.log(gardensBeforeSort);
    const newestThreeGardens: Kindergarten[] = fromArray.get3Max(MOCK_GARDENS);
    const gardensAfterSort = MOCK_GARDENS.map(g => g.id).join(' ');
    // console.log(gardensAfterSort);
    expect(newestThreeGardens).toHaveLength(3);
    expect(newestThreeGardens.map(g => g.id)).toEqual([2, 4, 7]);
    expect(gardensBeforeSort).toEqual(gardensAfterSort);
  });
});
