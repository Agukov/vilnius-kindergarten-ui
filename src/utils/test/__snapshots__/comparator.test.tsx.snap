// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compare functions should return 3 kidergardens with highset build date (ST) 1`] = `
Array [
  id: 1 build date: 2008-01-26T15:52:14+02:00 www: string,
  id: 2 build date: 2020-09-29T07:58:57+03:00 www: string,
  id: 3 build date: 2005-03-21T00:25:34+02:00 www: string,
  id: 4 build date: 2017-07-29T22:50:34+03:00 www: string,
  id: 5 build date: 2001-09-24T21:18:54+03:00 www: string,
  id: 6 build date: 2002-01-19T15:05:34+02:00 www: string,
  id: 7 build date: 2008-09-14T04:25:34+03:00 www: string,
  id: 8 build date: 1998-08-04T21:32:14+03:00 www: string,
  id: 9 build date: 2027-09-19T15:05:34+03:00 www: ,
]
`;

exports[`Compare functions should return 3 kidergardens with highset build date (ST) 2`] = `
Array [
  id: 2 build date: 2020-09-29T07:58:57+03:00 www: string,
  id: 4 build date: 2017-07-29T22:50:34+03:00 www: string,
  id: 7 build date: 2008-09-14T04:25:34+03:00 www: string,
]
`;

exports[`Compare functions should return 3 kidergardens with highset build date (ST) 3`] = `
Array [
  id: 1 build date: 2008-01-26T15:52:14+02:00 www: string,
  id: 2 build date: 2020-09-29T07:58:57+03:00 www: string,
  id: 3 build date: 2005-03-21T00:25:34+02:00 www: string,
  id: 4 build date: 2017-07-29T22:50:34+03:00 www: string,
  id: 5 build date: 2001-09-24T21:18:54+03:00 www: string,
  id: 6 build date: 2002-01-19T15:05:34+02:00 www: string,
  id: 7 build date: 2008-09-14T04:25:34+03:00 www: string,
  id: 8 build date: 1998-08-04T21:32:14+03:00 www: string,
  id: 9 build date: 2027-09-19T15:05:34+03:00 www: ,
]
`;
